Plot warning message

ColinTalbert opened the pull request:
This fixes the error message by changing the matplotlib backend to agg 
before any import of pyplot.

brendan-ward commented on tests/__init__.py:
```diff
@@ -1 +1,6 @@
 #
+try:
```
@ColinTalbert is there a reason for putting this here rather than what we already have in `test_tool.py`?

ColinTalbert replied:
@brendan-ward, It wasn't my first choice, but you need to change the backend to agg before any import of mpl.pyplot is done.  I originally tried it in test_tool.py but it didn't work (the plots still blocked)  I'm assuming it was because that import had already happened somewhere else by the time that test was run.

sgillies commented on rasterio/plot.py:
```diff
@@ -50,6 +50,9 @@ def show(source, cmap='gray', with_bounds=True):
     """
+    if plt is None:  # pragma: no cover
+        raise ImportError("Could not import matplotlib")
+
```
Instead of catching the `ImportError` at the top of the module and then reraising it here if `plt` is `None`, how about turning https://github.com/mapbox/rasterio/issues/741 into a `get_plt()` function? Call `get_plt()` at line 53 above and there's no need to check the value of `plt`.

sgillies replied:
In other words, while we prefer to keep our imports at the top of the file `matplotlib` is one situation where it's okay to import inside a function.

ColinTalbert replied:
I think using a "lazy import" so that matplotlib remains an optional dependency is an accepted design pattern. Moving this into a separate get_plt() function so it can be used from multiple plotting functions only makes sense. I guess I don't understand #741 enough to figure where it fits into that.

sgillies replied:
@ColinTalbert I made a cut and paste error above: I meant that the code of https://github.com/ColinTalbert/rasterio/blob/faf10ed7c04ee6410e92b0d545a05b1598c11617/rasterio/plot.py#L17-L28 could become `get_plt()`. Sorry for the confusion!

sgillies commented:
As for the right place to set the matplotlib backend: let's isolate this to individual tests capabilities of pytest, I think https://pytest.org/latest/monkeypatch.html could be handy. I think we could also refactor these tests to skip instead of passing if matplotlib isn't imported.

ColinTalbert commented:
So I think I'm all good here.   I implemented the get_plt() function.  I replaced the setting of the mpl backend call with an empty lambda function for plt.show(), and I set the tests to skip instead of pass if mpl is not installed.

Unfortunately I'm not sure how to correct the merge conflict from #745. I tried to cherry pick that commit but it would appear that I'm making more of a mess.  Any git guru's on the project that could provide assistance?

perrygeo commented:
@ColinTalbert I'm not sure exactly whats going on here (I would _not_ claim to be a git guru that's for sure!) but it looks like there are some commits from other work (docs and compat.py related).

The matplotlib stuff is isolated to 5 or 6 files so maybe the best course of action would be to manually apply those changes onto a clean branch from the current master. Sometimes that's just easier than fighting a messy merge.

perrygeo commented:
@ColinTalbert it looks like these changes are covered in #765, closing in favor of that PR
